File: iosgl/iosgl/bt/res/SceneRes.m
//
//  SceneRes.m
//  iosgl
//
//  Created by zhao on 12/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "SceneRes.h"
#import "ByteArray.h"
#import "Scene_data.h"
#import "LoadManager.h"
#import "GL_Header.h"

@interface SceneRes ()

@property(nonatomic,strong)SuccessBlock bfun;


@end
@implementation SceneRes
-(void)load:(NSString *)url bfun:(SuccessBlock)bfun;
{
    
    self.bfun=bfun;
     //本地文件读取
   /*
    NSString *path=  [[NSBundle mainBundle]pathForResource:@"5555_base" ofType:@"txt"];
    NSData *reader = [[NSData alloc] initWithContentsOfFile:path];
    NSLog(@"-----length----%lu",   reader.length);
    self.byte=[[ByteArray alloc]init:reader];
    [self loadComplete:self.byte];
  */

    
    [[LoadManager default] loadUrl:url type: LoadManager.IMG_TYPE fun:^(NSObject* value) {
        NSDictionary* bvalue=(NSDictionary*)value;
        NSData* netNsData = [[NSData alloc] initWithContentsOfFile:bvalue[@"data"]];
        self.byte=[[ByteArray alloc]init:netNsData];
        [self loadComplete:self.byte];
   
    }];
    
}
-(void)loadComplete:(ByteArray *)byte;
{
    [self applyByteArray];
    
}
-(void)applyByteArray;
{
    self.version = [self.byte readInt];
    NSLog(@"version-->%d", self.version);
    [self read]; //img
    [self read];//obj
    [self read];//material
    [self read];//particle
    [self readScene];
    
    self.bfun(nil);
}
-(void)readScene;
{
    int types = [self.byte readInt];
    [self readAstat];
    [self readTerrainIdInfoBitmapData:self.byte];
    int sceneInfosize   = [self.byte readInt];
    NSString *jsonStr= [self.byte readUTFBytes:sceneInfosize];
    NSData *data=[jsonStr dataUsingEncoding:NSUTF8StringEncoding];
    self.sceneData = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
}
-(void)readAstat;
{
    Boolean hasAstat = [self.byte readBoolean] ;
    if (hasAstat) {
        float midu = [self.byte readFloat];
        float aPosx = [self.byte readFloat];
        float aPosy = [self.byte readFloat];
        float aPosz = [self.byte readFloat];
        
        int tw = [self.byte readInt ];
        int th = [self.byte readInt ];
        
        NSLog(@"%d=%d",tw,th);
        
        float heightScaleNum= [self.byte readFloat];
        [self readAstarFromByte:self.byte];
        [self readAstarFromByte:self.byte];
        for (int i = 0; i < th; i++) {
            for (int j = 0; j < tw; j++) {
                [self.byte readShort ];
            }
            
        }
    }
    
}
-(void)readTerrainIdInfoBitmapData:(ByteArray *)byteSrc;
{
    int  len = [byteSrc readInt];
    if (len) {
        NSData *bmpNsdata=  [self.byte getNsDataByLen:len];
        ByteArray *materByte=  [[ByteArray alloc]init:bmpNsdata];
    }
    
}
 
-(void)readAstarFromByte:(ByteArray *)byteSrc
{
    int  len = [byteSrc readUnsignedInt];
    int intLen = ceil(((double)len) / 32.0);
    for (int i = 0; i < intLen; i++) {
        [byteSrc readUnsignedInt];
        for (int j = 0; j < 32; j++) {
            
        }
    }
    
}


@end

File: androldDemo/ab/app/src/main/java/z3d/res/SceneRes.java
package z3d.res;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import z3d.base.ByteArray;
import z3d.base.CallBackFun;


public class SceneRes extends BaseRes {

    public  JSONObject  sceneData;
    private CallBackFun sceneFinishFun;
    public void load(String url) {

    }
    public void  loadComplete(byte[] buff, CallBackFun bfun)
    {
        this.sceneFinishFun=bfun;
        this._byte =new ByteArray(buff);
        this.applyByteArray();
    }
    public void applyByteArray() {

        this.version = this._byte.readInt();
        Log.d("SceneRes版本 ->",   this.version+"===> " );
        this.read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext();
            }
        });
    }
    private void   readNext()
    {
        read();
        read();
        read();

        this.readScene();


    }
    private void  readScene()
    {
        int types = this._byte.readInt();
        this.readAstat();

        if (this.version >= 28) {
            this.readTerrainIdInfoBitmapData(this._byte);
        }
        int size   = this._byte.readInt();
        try {
            this.sceneData= new JSONObject(this._byte.readUTFBytes(size));

        } catch (JSONException e) {
            e.printStackTrace();
        }
        Log.d("partic地址 ->",   "aa");


        this.sceneFinishFun.StateChange(true);

    }
    private void readTerrainIdInfoBitmapData(ByteArray $byte) {
        int $len = $byte.readInt();
        if ($len>0) {
            byte[] objByte= $byte.readBytes($len);

        }

    }

    private void readAstat() {
        boolean hasAstat = this._byte.readBoolean();
        if (hasAstat) {

            float midu = this._byte.readFloat();
            float aPosx = this._byte.readFloat();
            float aPosy = this._byte.readFloat();
            float aPosz = this._byte.readFloat();
            int i ;
            int j;
            int tw = this._byte.readInt();
            int th = this._byte.readInt();

            if (this.version < 25) {
                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                        this._byte.readFloat();
                    }

                }
                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                        this._byte.readFloat();
                    }

                }
            } else {
               this._byte.readFloat();
                this.readAstarFromByte(this._byte);
                 this.readAstarFromByte(this._byte);



                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                         this._byte.readShort()  ;
                    }

                }

            }
        }
    }
    private void readAstarFromByte(ByteArray $byte){
        int $len = $byte.readUnsignedInt();
        int $intLen =(int) Math.ceil($len / 32.0f);
        for (int i = 0; i < $intLen; i++) {
             $byte.readUnsignedInt();

        }

    }



}
